File: Python/section1.3/transform.py
"""
ID: jameszh1
LANG: PYTHON2
TASK: transform
"""
def clockwise90(before):
    row = len(before)
    after = [([0 for i in range(row)]) for j in range(row)]
    for i in range(row):
        for j in range(row):
            after[j][row - 1 - i] = before[i][j]
    return after

def reflection(before):
    row = len(before)
    after = [([0 for i in range(row)]) for j in range(row)]
    for i in range(row):
        for j in range(row):
            after[i][j] = before[i][row - j - 1]
    return after

def check_matrix_equal(before ,after):
    row = len(before)
    for i in range(row):
        for j in range(row):
            if (before[i][j] != after[i][j]):
                return False
    return True

def is_clockwise90(before, after):
    return check_matrix_equal(clockwise90(before), after)

def is_clockwise180(before, after):
    return check_matrix_equal(clockwise90(clockwise90(before)), after)

def is_clockwise270(before, after):
    return check_matrix_equal(clockwise90(clockwise90(clockwise90(before))), after)

def is_reflection(before, after):
    return check_matrix_equal(reflection(before), after)

def is_combination(before, after):
    before = reflection(before)
    return is_clockwise90(before, after) or is_clockwise180(before, after) or is_clockwise270(before, after)

def is_noChange(before, after):
    row = len(before)
    for i in range(row):
        for j in range(row):
            if (before[i][j] != after[i][j]):
                return False
    return True

fin = open ('transform.in', 'r')
fout = open ('transform.out', 'w')
row = int(fin.readline().strip())
before = [([0 for i in range(row)]) for j in range(row)]
after = [([0 for i in range(row)]) for j in range(row)]
for i in range(row):
    line = fin.readline().strip()
    for j in range(row):
        before[i][j] = line[j]
for i in range(row):
    line = fin.readline().strip()
    for j in range(row):
        after[i][j] = line[j]

if (is_clockwise90(before, after)):
    fout.write("1" + '\n')
elif (is_clockwise180(before, after)):
    fout.write("2" + '\n')
elif (is_clockwise270(before, after)):
    fout.write("3" + '\n')
elif (is_reflection(before, after)):
    fout.write("4" + '\n')
elif (is_combination(before, after)):
    fout.write("5" + '\n')
elif (is_noChange(before, after)):
    fout.write("6" + '\n')
else:
    fout.write("7" + '\n')
fout.close()

File: Java/section1.3/transform.java
/*
ID: jameszh1
LANG: JAVA
TASK: transform
*/
import java.io.*;
public class transform {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new FileReader("transform.in"));
        PrintWriter pw = new PrintWriter(new BufferedWriter(new FileWriter("transform.out")));
        //initialization
        int n = Integer.parseInt(br.readLine());
        int[][] before = new int[n][n];
        int[][] after = new int[n][n];
        for (int i = 0; i < n; i++) {
            String line = br.readLine();
            for (int j = 0; j < line.length(); j++) {
                before[i][j] = line.charAt(j);
            }
        }
        for (int i = 0; i < n; i++) {
            String line = br.readLine();
            for (int j = 0; j < line.length(); j++) {
                after[i][j] = line.charAt(j);
            }
        }

        if (clockwise90(before, after)) pw.println("1");
        else if (clockwise180(before, after)) pw.println("2");
        else if (clockwise270(before, after)) pw.println("3");
        else if (reflection(before, after)) pw.println("4");
        else if (combination(before, after)) pw.println("5");
        else if (noChange(before, after)) pw.println("6");
        else  pw.println("7");


        pw.close();
        br.close();
    }

    public static boolean clockwise90(int[][] before, int[][] after) {
        boolean flag = true;
        int n = before.length;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (before[i][j] != after[j][n - 1 - i]) return false;
            }
        }
        return flag;
    }

    public static boolean clockwise180(int[][] before, int[][] after) {
        boolean flag = true;
        int n = before.length;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (before[i][j] != after[n - 1 - i][n - 1 - j]) return false;
            }
        }
        return flag;
    }

    public static boolean clockwise270(int[][] before, int[][] after) {
        boolean flag = true;
        int n = before.length;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (before[i][j] != after[n- 1- j][i]) return false;
            }
        }
        return flag;
    }

    public static boolean reflection(int[][] before, int[][] after) {
        boolean flag = true;
        int n = before.length;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (before[i][j] != after[i][n - 1 - j]) return false;
            }
        }
        return flag;
    }

    public static boolean combination(int[][] before, int[][] after) {
        boolean flag = true;
        int n = before.length;
        int[][] newBefore = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                newBefore[i][j] = before[i][n - 1 - j];
            }
        }
        if (clockwise90(newBefore, after) || clockwise180(newBefore, after) || clockwise270(newBefore, after)) {
            return true;
        }
        else {
            flag = false;
        }
        return flag;
    }

    public static boolean noChange (int[][] before, int[][] after) {
        boolean flag = true;
        int n = before.length;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (before[i][j] != after[i][j]) return false;
            }
        }
        return flag;
    }
}
